### ISSUE TITLE
Option to suppress "Unknown format" errors

### ISSUE THREAD

**Issue by rbren** [ajv-validator/ajv#466]
I'm using AJV in conjunction with a bunch of OpenAPI specs, which allow anything to be in the "format" field.  With `unknownFormats` set to `ignore`, I get a wall of error messages.

**What version of Ajv you are you using?**
4.11.5

**What problem do you want to solve?**
Suppress "unknown format" errors

**What do you think is the correct solution to problem?**
Add a new option ("suppress")

**Will you be able to implement it?**
Would be happy to, if you could point me in the right direction

**Comment by epoberezkin** [comment]
Not sure I understand why it is needed. The ability to quietly suppress all unknown format will lead to mistakes and unnoticed typos (as false positive results are usually the most difficult to detect with JSON schema).

Although the value of format can be any string, it has some semantic meaning, so your options currently are:

- fail on unknown formats - you will need to add them all, and you can always add them as `ajv.addFormat('name', /.*/)` if you want to delay the actual implementation.
- whitelist unknown formats you want to ignore by passing an array of values to unknownFormats option. In this case you won't get warnings if these formats are used.
- "ignore" all unknown formats (which is prone to errors even with warnings).

I don't understand why you need an additional value for this option and whitelisting those formats you don't want to implement is not good enough.

What is the use case?

Also, the latest version is 5.0.0 - https://github.com/epoberezkin/ajv/releases/tag/5.0.0

**Comment by epoberezkin** [comment]
Also if you want to disable formats completely you can use option `format: false`, in which case ALL formats will be quietly ignored

**Comment by rbren** [comment]
I'm working with a constantly changing list of schemas from the wild, so I can't whitelist. I'd like to *completely* ignore unknown formats, but validate the ones supported by JSON schema.

`format: false` helps, but I'm guessing that disables e.g. email validation.

**Comment by epoberezkin** [comment]
> I'm guessing that disables e.g. email validation

That is correct.

I am still not convinced that suppressing warning on a library level is a good idea... The problem here, even in your case, is that you are using third party schemas ignoring some part of them without providing any feedback to providers of those schemas. An option for your app/library here would be to process and report these warnings somehow.

If anything, there should either be an option to suppress all warnings (e.g. quiet: true) or for more granular control all options that may lead to warning should support value "quiet". Or both...

**Comment by epoberezkin** [comment]
#618 should address it - it will be possible to capture all warnings/errors in a function and to decide there whether to log them.

**Comment by alirobe** [comment]
Because this came up while googling... 

According to this book, validators should ignore formats that aren't understood by default, so I was initially confused by this behaviour: 
https://json-schema.org/understanding-json-schema/reference/string.html#format

But actually the docs are pretty clear on this. That's not AJV's default way of dealing with unknown formats:
https://ajv.js.org/#formats

The option requested here now exists, just set "unknownFormats" to "ignore", or ["myformata", "myformatb"]

**Comment by epoberezkin** [comment]
The problem with simply ignoring anything is that you would not catch typos and you never know what is ignored. That’s the reason why default is not to ignore. Permissiveness of JSON Schema makes it very difficult to maintain schemas - can you imagine a programming language that ignores all statements and that it doesn’t know about? Well, it’s JSON Schema.

You can also whitelist formats that may be used in schema and not supported

**Comment by jorisw** [comment]
I cannot get my schema with a custom format to validate.

```js
const schema = {
  postcode: {
    title: 'form.address.postcode',
    type: 'string',
    format: 'postcode'
  },
  houseNumber: { title: 'form.address.house-number', type: [ 'string' ] },
  extension: { title: 'form.address.extension', type: [ 'string' ] },
  addressDetails: { title: 'form.address.your-address', type: 'string' },
  additionalProperties: false,
}

const validator = new Ajv({ unknownFormats: 'ignore' })
const validate = validator.compile(schema)

const formInput = { postcode: "1234AB" }

const valid = validate(formInput)
```

Throws:

`unknown format "postcode" ignored in schema at path "#/properties/postcode"`
